www/caddy/src/opnsense/mvc/app/views/OPNsense/Caddy/diagnostics.volt

```diff
@@ -0,0 +1,79 @@
+{#
+ # Copyright (c) 2024 Cedrik Pischem
+ # All rights reserved.
+ #
+ # Redistribution and use in source and binary forms, with or without modification,
+ # are permitted provided that the following conditions are met:
+ #
+ # 1. Redistributions of source code must retain the above copyright notice,
+ #    this list of conditions and the following disclaimer.
+ #
+ # 2. Redistributions in binary form must reproduce the above copyright notice,
+ #    this list of conditions and the following disclaimer in the documentation
+ #    and/or other materials provided with the distribution.
+ #
+ # THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ # INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+ # AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ # AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ # OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ # SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ # INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ # POSSIBILITY OF SUCH DAMAGE.
+ #}
+
+<script type="text/javascript">
+    $(document).ready(function() {
+        // Download the json configuration as file with timestamp
+        $("#downloadCaddyConfig").click(function() {
+            $.ajax({
+                url: "/api/caddy/diagnostics/showconfig",  // Custom API endpoint that shows a validated json configuration
+                type: "GET",
+                success: function(response) {
+                    if (response.status === "success") {
+                        let timestamp = new Date().toISOString().replace(/[-:]/g, '').replace('T', '-').split('.')[0];
+                        let filename = "caddy_autosave_" + timestamp + ".json";
+                        download_content(JSON.stringify(response.content), filename, "application/json");
+                    } else {
+                        alert("Failed to download configuration: " + response.message);
+                    }
+                },
+                error: function() {
+                    alert("Error accessing the API.");
+                }
+            });
+        });
+
+        // Open the json configuration in a new browser tab
+        $("#openConfigInNewTab").click(function() {
+            window.open('/api/caddy/diagnostics/showconfig', '_blank');
```

Comment:
Unsure if this is allowed.

Reply:
why not show it in a modal?

Reply:
Since browser have great built in json viewers, maintaining a modal for that would be kind of a downgrade.

Reply:
Isn’t that making the download more or less redundant then if you let the browser view a flat file anyway?

Reply:
Refactored it to show everything in the UI.